### PR TITLE
Adding http Request/Response injection

### PR DESCRIPTION
**Pull request by logangirvin**
**Problem:**
Current Wallet library offers no means to inject custom request headers, time requests, or perform more advanced logging.

**Solution:**
Adds an abstraction for wallet developers to hook into requests before being sent and time responses.

**Validation:**
manually integrated testing

**Type of change:**
- [x] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

**Risk**:
- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [x] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

**Work Item links**:
Please include here links for this work item, or deferred work, or related work. E.g. if the refactoring is too big to fit in this PR, or the localized strings need to be updated later, please link the TODO work items here.

**Documentation Links**:
Please include here links to any related background documentation for this PR.

### CODE REVIEW

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/interceptors/ExternalInterceptor.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            headers = this.convertHeaders(originalRequest.headers)
+        )
+
+        // allow each interceptor to mutate the headers
+        httpInterceptors.forEach {
+            it.beforeRequest(externalRequest)
+        }
+        // build a new request if required
+        val newBuilder = originalRequest.newBuilder()
+        var changed = false
```
Comment (symorton) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/interceptors/ExternalInterceptor.kt:34]: why does changed need to be here?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/VerifiableCredentialSdk.kt [hunk 2]
```diff
@@ -69,7 +71,8 @@ internal object VerifiableCredentialSdk {
         polymorphicJsonSerializers: SerializersModule = Json.serializersModule,
         registrationUrl: String = "",
         resolverUrl: String = "https://discover.did.msidentity.com/v1.0/identifiers",
-        walletLibraryVersionInfo: String = ""
+        walletLibraryVersionInfo: String = "",
+        interceptors: List<HttpInterceptor> = emptyList()
```
Comment (symorton) [inline walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/VerifiableCredentialSdk.kt:75]: I think this is fine for now, but will be ever want to allow the developers to only add an httpInterceptor to certain network call. Does this allow for that? Also, are there any security implications with these interceptors?

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/interceptors/ExternalInterceptor.kt (hunk 1 above)
Comment (logangirvin) [reply walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/interceptors/ExternalInterceptor.kt:34]: Attempting to reduce rebuilding the request if the interceptor isn't used or doesn't modify headers.

File: walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/VerifiableCredentialSdk.kt (hunk 2 above)
Comment (logangirvin) [reply walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/VerifiableCredentialSdk.kt:75]: Sort of, thats why parameters are passed to identify the uri, method, and headers, but it does put the bulk of the request identifying work on developers. 
I also realized it should probably be a base class with no-op implementations so you only implement/override what you need.

**Pull request closed without merging**